Invariant EKF

dellaert opened the pull request:
A Left Invariant IEKF on SE(2).
Changes by Frank: template methods rather than a base class
Replaces #2113 

This introduced the following class structure:
```mermaid
graph TD; 
ManifoldEKF --> LieGroupEKF --> InvariantEKF
````

scottiyio committed:
Add files via upload

A Left Invariant IEKF on SE(2).

scottiyio committed:
Add files via upload

scottiyio committed:
Add files via upload

scottiyio committed:
Add files via upload

scottiyio committed:
Delete examples/SE2LIEKF_NoFactors.cpp

scottiyio committed:
clang-formatted, changed getFoobar() to foobar()

scottiyio committed:
Added documentation and format changes. 

- Added comments
- Clang-formatted

scottiyio committed:
Added authors

scottiyio committed:
Added documentation & clang formatting

scottiyio committed:
Added authors

dellaert committed:
no_boost

dellaert committed:
Templated predict/update

dellaert committed:
More symmetric templating

dellaert committed:
Comments with o4-mini

dellaert committed:
Move to nav

dellaert committed:
Docs with o4-mini

dellaert committed:
Test WIP

dellaert committed:
Use nav version

dellaert committed:
Added Rot3 example to show state-dependent control

dellaert commented:
Ci on windows fails because of Unicode characters. I’ll fix together with any other comments.

scottiyio commented on examples/LIEKF_NavstateExample.cpp:
```diff
@@ -0,0 +1,96 @@
+ */
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/OptionalJacobian.h>
```
is this needed since OptionalJacobian.h is included in LIEKF.h?

scottiyio commented on examples/LIEKF_NavstateExample.cpp:
```diff
@@ -0,0 +1,96 @@
+ * @authors Scott Baker, Matt Kielo, Frank Dellaert
+ */
+
+#include <gtsam/base/Matrix.h>
```
are line 19 and 20 needed includes, since they are already included in LIEKF.h?

scottiyio commented on examples/LIEKF_Rot3Example.cpp:
```diff
@@ -0,0 +1,78 @@
+ */
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/OptionalJacobian.h>
```
same includes - required or not?

scottiyio commented on examples/LIEKF_SE2_SimpleGPSExample.cpp:
```diff
@@ -0,0 +1,118 @@
+  LIEKF<Pose2> ekf(X0, P0);
+
+  // Define the process covariance and measurement covariance matrices Q and R.
+  Matrix3 Q = (Vector3(0.05, 0.05, 0.001)).asDiagonal();
```
Is there any better way to initialize matrices like this when we want different sigma^2 on the diagonal? Or is this just the best way if they are different valurd?

scottiyio commented on examples/LIEKF_SE2_SimpleGPSExample.cpp:
```diff
@@ -0,0 +1,118 @@
+
+  // Define the process covariance and measurement covariance matrices Q and R.
+  Matrix3 Q = (Vector3(0.05, 0.05, 0.001)).asDiagonal();
+  Matrix2 R = (Vector2(0.01, 0.01)).asDiagonal();
```
change to Matrix2::Identity() * 0.01

scottiyio replied:
May want to find a better way to create covariance matrices when the sigma^2 are different on the diagonal

scottiyio commented on examples/LIEKF_NavstateExample.cpp:
```diff
@@ -0,0 +1,96 @@
+ */
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/OptionalJacobian.h>
```
Line (19) and (20) are already included in LIEKF.h

dellaert replied:
I try to follow https://google.github.io/styleguide/cppguide.html#Include_What_You_Use

scottiyio commented on examples/LIEKF_Rot3Example.cpp:
```diff
@@ -0,0 +1,78 @@
+ */
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/OptionalJacobian.h>
```
Line (20) and (21) are already included in LIEKF.h

dellaert replied:
Same: IWYU

scottiyio commented on gtsam/navigation/LIEKF.h:
```diff
@@ -0,0 +1,185 @@
+#pragma once
+
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/OptionalJacobian.h>
```
Same includes are not needed, in LIEKF.h already

scottiyio reviewed (approved):
Looks great! I saw no errors, just some minor includes that may not be needed.

dellaert committed:
Review comments

dellaert committed:
SFINAE to restrict template matching

dellaert committed:
Update testLIEKF.cpp

dellaert committed:
well-tested predictMean

dellaert committed:
predictMean for state=control f

dellaert committed:
Correct jacobian for h_gps

dellaert committed:
Rename

dellaert committed:
Fix compilation

dellaert commented:
@scottiyio I made the following changes:
- I added new unit-tests for state-dependent dynamics, with new method `predictMean` 
- renamed to InvariantEKF
- reverted this to draft PR because the connection with invariant filtering is not clear enough yet

I would like to look at Potokar’s code how the dynamics and measurements there are specified in general.

dellaert committed:
Add value to concepts

dellaert committed:
EKF hierarchy

dellaert commented:
@scottiyio Gemini and I created the ManifoldEKF -> GroupEKF -> InvariantEKF hierarchy (plus tests)
Take a look.
GroupEKF or LieEKF or LieGroupEKF?

dellaert commented on examples/IEKF_NavstateExample.cpp:
```diff
@@ -0,0 +1,95 @@
+ * @param imu  6×1 vector [a; ω]: linear acceleration and angular velocity.
+ * @return     9×1 tangent: [ω; 0₃; a].
+ */
+Vector9 dynamics(const Vector6& imu) {
```
@scottiyio I think this example still needs some unit-testing. In particular, show that
- $X \cdot U = X \cdot \exp(\delta t \xi^\wedge)$ the same as the nonlinear dynamics in Potokar
- Show that $Ad_{U^{-1}}$ matches the LIEKF Jacobian in Potokar.

scottiyio replied:
I'll write some test cases here

dellaert commented on gtsam/navigation/GroupEKF.h:
```diff
@@ -0,0 +1,196 @@
+      static_assert(IsLieGroup<G>::value, "Template parameter G must be a GTSAM Lie Group");
+    }
+
+    /**
```
I am thinking of *moving* these to InvariantEKF and comment that, if the dynamics does not depend on the state, they should use an InvariantEKF

scottiyio replied:
I know that InvariantEKF inherits from this.

I wish we could call it something like "StateDependentEKF" because you should use GroupEKF on state dependency, but considering InvariantEKF inherits from here as a special case I don't know if that makes sense.

Or maybe, it inherits that the "state dependency" is zero. That structure, to me, makes it more clear thaty you should use Invariant if you have no state dependency but Group if there is.

dellaert replied:
I opted for
```mermaid
graph TD; 
ManifoldEKF --> LieGroupEKF --> InvariantEKF
````

dellaert committed:
Renamed GroupEKF -> LieGroupEKF, moved invariant predicts
